Origin: WMF
Description: introduces a new optional flag (-f) to vcl.discard. If the flag is
specified, immediately discard the given VCL
Index: varnish4/bin/varnishd/cache/cache_vcl.c
===================================================================
--- varnish4.orig/bin/varnishd/cache/cache_vcl.c
+++ varnish4/bin/varnishd/cache/cache_vcl.c
@@ -180,6 +180,19 @@ vcl_find(const char *name)
 	return (NULL);
 }
 
+static struct vcl *
+vcl_find_discarded_is_fine_too(const char *name)
+{
+       struct vcl *vcl;
+
+       ASSERT_CLI();
+       VTAILQ_FOREACH(vcl, &vcl_head, list) {
+               if (!strcmp(vcl->loaded_name, name))
+                       return (vcl);
+       }
+       return (NULL);
+}
+
 /*--------------------------------------------------------------------*/
 
 void
@@ -893,7 +906,29 @@ vcl_cli_discard(struct cli *cli, const char * const *av, void *priv)
 	ASSERT_CLI();
 	(void)cli;
 	AZ(priv);
-	vcl = vcl_find(av[2]);
+	if (!strcmp(av[2], "-f") && av[3] == NULL) {
+		VCLI_Out(cli, "Too few parameters");
+		VCLI_SetResult(cli, CLIS_TOOFEW);
+		return;
+	} else if (strcmp(av[2], "-f") && av[3] != NULL) {
+		VCLI_Out(cli, "Unknown options '%s'", av[2]);
+		VCLI_SetResult(cli, CLIS_PARAM);
+		return;
+	} else if (av[3] != NULL) {
+		vcl = vcl_find_discarded_is_fine_too(av[3]);
+		if (vcl == NULL) {
+			return;
+		}
+
+		if (vcl->discard) {
+			// the admin is forcing removal of an already discarded VCL. It must be stuck!
+			vcl->temp = VCL_TEMP_COLD;
+			vcl->busy = 0;
+			return;
+		}
+	} else
+		vcl = vcl_find(av[2]);
+
 	AN(vcl);			// MGT ensures this
 	Lck_Lock(&vcl_mtx);
 	assert (vcl != vcl_active);	// MGT ensures this
Index: varnish4/include/tbl/cli_cmds.h
===================================================================
--- varnish4.orig/include/tbl/cli_cmds.h
+++ varnish4/include/tbl/cli_cmds.h
@@ -94,10 +94,10 @@ CLI_CMD(VCL_STATE,
 
 CLI_CMD(VCL_DISCARD,
 	"vcl.discard",
-	"vcl.discard <configname|label>",
+	"vcl.discard [-f] <configname|label>",
 	"Unload the named configuration (when possible).",
 	"",
-	1, 1
+	1, 2
 )
 
 CLI_CMD(VCL_LIST,
Index: varnish4/bin/varnishd/mgt/mgt_vcl.c
===================================================================
--- varnish4.orig/bin/varnishd/mgt/mgt_vcl.c
+++ varnish4/bin/varnishd/mgt/mgt_vcl.c
@@ -699,6 +699,14 @@ mcf_vcl_discard(struct cli *cli, const c
 	int n;
 
 	(void)priv;
+
+	if (av[3] != NULL) {
+		// The admin wants to force discard, skip all checks
+		(void)mgt_cli_askchild(&status, &p, "vcl.discard -f %s\n", av[3]);
+		free(p);
+		return;
+	}
+
 	vp = mcf_find_vcl(cli, av[2]);
 	if (vp == NULL)
 		return;
